text/0000-inline-semantic.md

```diff
@@ -0,0 +1,993 @@
+}
+fn main() {
+    println!("args[1] = {}", my_unwrap(args().nth(1)));
+    println!("args[2] = {}", my_unwrap(args().nth(2)));
+    println!("args[3] = {}", my_unwrap(args().nth(3)));
+}
+```
+
+This trivial implementation, however, will only report the panic happens inside `my_unwrap`. This is
+pretty useless, since it is the caller of `my_unwrap` that have made the wrong assumption!
+
+```text
+$ ./2
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1
+args[1] = arg1
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1 arg2
+args[1] = arg1
+args[2] = arg2
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1 arg2 arg3
+args[1] = arg1
+args[2] = arg2
+args[3] = arg3
+```
+
+The trivial solution would be requiring user to provide `file!()`, `line!()` and `column!()`. A
+slightly more ergonomic solution would be changing `my_unwrap` to a macro, thus these constants can
+be automatically provided.
+
+```rust
+macro_rules! my_unwrap {
+    ($input:expr) => {
+        match $input {
+            Some(t) => t,
+            None => panic!("nothing to see at {}:{}:{}, move along", file!(), line!(), column!()),
+        }
+    }
+}
+println!("args[1] = {}", my_unwrap!(args().nth(1)));
+//                                ^ tell user to add an `!`.
+...
+```
+
+But what if you have already published the `my_unwrap` crate that has thousands of users, and you
+want to maintain API stability? Before Rust 1.XX, the builtin `unwrap()` has the same problem!
+
+## Semantic-inlining
+
+The reason a `my_unwrap!` macro works is because it copy-and-paste the entire content of its macro
+definition everytime it is used.
+
+```rust
+println!("args[1] = {}", my_unwrap!(args().nth(1)));
+println!("args[2] = {}", my_unwrap!(args().nth(2)));
+...
+
+// is equivalent to:
+
+println!("args[1] = {}", match args().nth(1) {
+    Some(t) => t,
+    None => panic!("nothing to see at {}:{}:{}, move along", file!(), line!(), column!()),
+});
+println!("args[2] = {}", match args().nth(2) {
+    Some(t) => t,
+    None => panic!("nothing to see at {}:{}:{}, move along", file!(), line!(), column!()),
+});
+...
+```
+
+What if we allow *normal functions* be able to copy-and-paste its content as well? This is called
+**[inlining]**, which Rust and many other languages have supported this as an optimization step. If
+the function is inlined, the compiler can also tell the inlined function the location it is inlined,
+and thus able to report this to the user.
+
+Rust allows developers to use the `#[inline]` attribute to *hint* the optimizer that a function
+should be inlined. However, if we want the precise caller location, a hint is not enough, it needs
+to be a requirement. Therefore, the `#[inline(semantic)]` attribute is introduced.
```

Comment:
I think the description here is slightly off - we already have `#[inline(always)]`, and `#[inline(semantic)]` behaves fundamentally differently than traditional inlining. This feature doesn't even necessarily need to be thought about as an inlining at all, really - you could also phrase it in terms of parameterizing the function over its invocation locations, for example.